sphinx/registry.py

```diff
@@ -341,6 +347,16 @@ def add_autodoc_attrgetter(self, typ, attrgetter):
         # type: (Type, Callable[[Any, unicode, Any], Any]) -> None
         self.autodoc_attrgettrs[typ] = attrgetter
 
+    def add_javascript(self, filename):
```

Comment:
Why do we store this stuff in `Registry`? I thought this object was for global configuration, but scripts and stylesheets only apply to the HTML builders. Couldn't we keep storing this as an attribute of the HTML builders but simply read from the configuration value instead?

Reply:
> Note:
> Personally, app.add_javascript() and app.add_documenter() APIs are also not so good. Because both HTML builder and autodoc are not core components. They are only one of builder (or extension) for Sphinx. so application object should not provide APIs for individual sub-components.
> So these should be changed in future (if there is good idea!)

Yes, this is my concern **exactly**. It would seem we need a way for extensions to override configuration values that are not related to the `sphinx.application.Application` object. We could allow them to override various values read in from `conf.py` but that will create all types of chaos if done incorrectly (i.e. a misbehaving extension could completely override user-provided configuration). It would seem we are stuck with this API for now.

Reply:
No, your understanding is not right. These APIs does not related with users configurations basically. They are used to store "components" for Sphinx only. For example, builders, domains, parsers, and so on. JS and CSS files are one of them.

Certainly `add_javascript()` is related with `html_javascripts` which I added in this PR. This is an exception of rules above. `html_javascripts` is a configuration to register JS files to HTML builders. So it also registers "components".
FYI: If both are used at once, all javascripts are enabled, not overrided.


BTW, what I meant at last comment is that sphinx-core is a kind of framework. And HTML builder is one of its component. So sphinx-core should not treat sub-components (like JS and CSS) of these components directly. IMO, it is a responsibility of each component, not application's.

Reply:
FYI: The extensions can override the users' configuration on runtime if they want.
```
def on_config_inited(app, config):
    config.some_value = 1
    config.other_value = 2

def setup(app):
    app.connect('config-inited', on_config_inited)
```

Reply:
> No, your understanding is not right. These APIs does not related with users configurations basically. They are used to store "components" for Sphinx only. For example, builders, domains, parsers, and so on. JS and CSS files are one of them.

Sorry, when I said "them" I meant extension developers - not users.

> BTW, what I meant at last comment is that sphinx-core is a kind of framework. And HTML builder is one of its component. So sphinx-core should not treat sub-components (like JS and CSS) of these components directly. IMO, it is a responsibility of each component, not application's.

Yup, I was actually agreeing with you. It seems I was not clear enough 😄

Reply:
Okay. Anyway this PR does not change our API at present. We already provide APIs for adding JS and CSS, so this does not change public APIs.

So let's discuss about better APIs in other issue. As I commented above, I always welcome better interfaces.